### ISSUE TITLE
Simplify AdSense setup flow implementation

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#1014]
The AdSense module, particularly its setup implementation, has considerably outgrown its architecture in scope of which corner-cases it should cover. As such, the codebase is incredibly error-prone and hard to follow, which has resulted in several regressions in the past few months from even small changes.

The setup flow logic needs to be simplified.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a clear list of checks for where the AdSense account and site are, in individual functions.
* Every resulting screen should be its own individual component (reusing child components where possible of course).
* One central component should based on the input result from the checks decide which screen should be displayed.

## Implementation Brief

* Introduce internal AdSense settings (eg. flags) for `accountSetupComplete` and `siteSetupComplete`. These are flags that need to be set to indicate that the user has completed the setup parts for both pieces. The plain existence of an account ID, client ID, or something else does not provide that information. The settings are needed to indicate to both the server and the existing (rather confusing) JS logic to render the setup instead of the module dashboard.
    * Remove the current (too generic) `setupComplete` setting once the above two cover all instances.
* Add `settings.js` piece to `modules/adsense` datastore:
    * `submitChanges` action
    * `canSubmitChanges` selector
    * `isDoingSubmitChanges` selector
* Implement common AdSense components:
    * `AccountSelect` (for the edge-case of people having more than one account)
    * `UseSnippetSwitch` (with "Recommended" indicator and warning based on on/off status, initial value of on/off status, and whether there is an existing tag)
* Implement common (overall) components:
    * `SettingsNotice` (used in multiple places by AdSense module only, but actually generic)
* Implement AdSense setup components:
    * `SetupMain`
        * access data (mainly from #1332) via React hooks and then render sub component based on status (see below)
    * `SetupAccountCreate` (considers existing tag)
    * `SetupAccountSelect`
    * `SetupAccountDisapproved`
    * `SetupAccountPending` (used for both "pending" and "graylisted" status)
    * `SetupAccountNoClient`
    * `SetupAccountApproved` (only needs to be rendered when `accountSetupComplete` is not yet true; considers existing tag)
    * `SetupSiteAdd`
    * `SetupSiteAdded` (only needs to be rendered when `siteSetupComplete` is not yet true)
* Implement AdSense settings components:
    * `SettingsMain`
    * `SettingsEdit`
        * show `UseSnippetSwitch`
    * `SettingsView`
        * show same data as is displayed today (see #585)
* Hook `SetupMain` and `SettingsMain` up with current setup and settings filters, wrapped in a `Data.RegistryProvider` with our registry

## Changelog entry

* Significantly improve stability and maintainability of AdSense module setup and settings.

**Comment by tofumatt** [comment]
IB ✅

Looks good and seems complete to me. 👍

**Comment by cole10up** [comment]
**Tested**

Setup installed and activated latest develop version of SK. Activated Adsense for the following scenarios:

### Scenario: Adsense account but site not added:
![image](https://user-images.githubusercontent.com/57115907/82631845-65ed8800-9bab-11ea-8224-317a4191e21f.png)

Confirmed drives user to the correct URL:
![image](https://user-images.githubusercontent.com/57115907/82631862-730a7700-9bab-11ea-93df-03a3eab3d0a0.png)

Added the site to my account - confirmed SK now shows Adsense connected:
![image](https://user-images.githubusercontent.com/57115907/82631939-ae0caa80-9bab-11ea-9421-b9b347c1dc22.png)

![image](https://user-images.githubusercontent.com/57115907/82631950-b7961280-9bab-11ea-8171-ec514f71f5de.png)

### Scenario: Admob Account versus Adsense
![image](https://user-images.githubusercontent.com/57115907/82632282-a1d51d00-9bac-11ea-9a54-f0dcc591a6ff.png)

Confirmed link drives to:
![image](https://user-images.githubusercontent.com/57115907/82632309-b1ecfc80-9bac-11ea-9c08-9d0d8598ffe3.png)

Cancel and retry module setup functions properly

### Scenario: Disapproved account
![image](https://user-images.githubusercontent.com/57115907/82632499-3fc8e780-9bad-11ea-9e33-6c3d550608f1.png)

Clicking the link drives to:
![image](https://user-images.githubusercontent.com/57115907/82632519-4e170380-9bad-11ea-886a-9631b2043418.png)

Cancel and retry module setup functions properly

### Scenario: No Adsense account
![image](https://user-images.githubusercontent.com/57115907/82632925-94b92d80-9bae-11ea-8933-c77f7e7bda2e.png)

Clicking Learn more drives to:
![image](https://user-images.githubusercontent.com/57115907/82632953-a8649400-9bae-11ea-8cce-8bd32059fa92.png)

Clicking Create Adsense Account drives user here:
![image](https://user-images.githubusercontent.com/57115907/82633040-ed88c600-9bae-11ea-93df-a3fec0827012.png)

After selecting account user is driven to sign up page:
![image](https://user-images.githubusercontent.com/57115907/82633049-fb3e4b80-9bae-11ea-9ba1-b56061a73467.png)

Finishing setup:
![image](https://user-images.githubusercontent.com/57115907/82633087-20cb5500-9baf-11ea-960c-8fe9a7d82d37.png)

Returning to SK:
![image](https://user-images.githubusercontent.com/57115907/82633103-27f26300-9baf-11ea-94f8-d6572c23dc04.png)

Confirmed settings shows connected properly

**One thing to note - for consistency, user's setup page doesn't have the toggle to remove the Adsense code**

![image](https://user-images.githubusercontent.com/57115907/82633188-5a03c500-9baf-11ea-93b0-bdac3e494a4b.png)

@felixarntz - Can you confirm this is ok? Do we need a separate ticket for this?

### Scenario Adsense account and added with site added and revenue
![image](https://user-images.githubusercontent.com/57115907/82633340-b961d500-9baf-11ea-905e-8e6f0c055aae.png)

![image](https://user-images.githubusercontent.com/57115907/82633346-bff04c80-9baf-11ea-9542-d3c48d3fa78e.png)

Confirmed earnings are displayed:
![image](https://user-images.githubusercontent.com/57115907/82633373-ca124b00-9baf-11ea-9526-f7caa7b4f655.png)

![image](https://user-images.githubusercontent.com/57115907/82633401-dd251b00-9baf-11ea-8a91-3529b10fe6f7.png)

![image](https://user-images.githubusercontent.com/57115907/82633414-e4e4bf80-9baf-11ea-859c-8f61b4af888a.png)

Confirmed settings page:
![image](https://user-images.githubusercontent.com/57115907/82633434-f201ae80-9baf-11ea-88da-bfb01e0a3332.png)

Checked status:
![image](https://user-images.githubusercontent.com/57115907/82633449-f8902600-9baf-11ea-829e-d3931140b803.png)

Editing status:
![image](https://user-images.githubusercontent.com/57115907/82633468-034abb00-9bb0-11ea-98dc-6d37bc95f610.png)

Confirmed disconnect functions:
![image](https://user-images.githubusercontent.com/57115907/82633479-09409c00-9bb0-11ea-9edb-89c9d3d9ac31.png)

Passed QA ✅
